prusti-tests/tests/compiletest.rs

```diff
@@ -0,0 +1,184 @@
+extern crate prusti_server;
+
+use compiletest_rs::{common, run_tests, Config};
+use prusti_server::ServerSideService;
+use std::{env, path::PathBuf};
+
+fn get_prusti_rustc_path() -> PathBuf {
+    let local_prusti_rustc_path: PathBuf = if cfg!(windows) {
+        ["target", "debug", "prusti-rustc.exe"].iter().collect()
+    } else {
+        ["target", "debug", "prusti-rustc"].iter().collect()
+    };
+    let workspace_prusti_rustc_path: PathBuf = if cfg!(windows) {
+        ["..", "target", "debug", "prusti-rustc.exe"]
+            .iter()
+            .collect()
+    } else {
+        ["..", "target", "debug", "prusti-rustc"].iter().collect()
+    };
+    if local_prusti_rustc_path.exists() {
+        return local_prusti_rustc_path;
+    }
+    if workspace_prusti_rustc_path.exists() {
+        return workspace_prusti_rustc_path;
+    }
+    panic!("Could not find the prusti-rustc binary to be used in tests");
+}
+
+/// Set an environment variable, automatically reverting the change at the end of the lexical scope
+struct TemporaryEnvVar {
+    name: String,
+    original: Option<String>,
+}
+
+impl TemporaryEnvVar {
+    fn set(name: &str, value: &str) -> Self {
+        let original: Option<String> = env::var(name).ok();
+        env::set_var(name, value);
+        TemporaryEnvVar {
+            name: name.to_string(),
+            original,
+        }
+    }
+}
+
+impl Drop for TemporaryEnvVar {
+    fn drop(&mut self) {
+        match self.original {
+            Some(ref old_value) => env::set_var(&self.name, old_value),
+            None => env::remove_var(&self.name),
+        }
+    }
+}
+
+fn run_prusti_tests(group_name: &str, filter: &Option<String>, rustc_flags: Option<&str>) {
+    let mut config = Config::default();
+    config.rustc_path = get_prusti_rustc_path();
+
+    // Filter the tests to run
+    config.filter = filter.clone();
+
+    // Add compilation flags
+    config.target_rustcflags = Some(format!(
+        "--edition=2018 {}",
+        rustc_flags.unwrap_or("")
+    ));
+
+    let path: PathBuf = ["tests", group_name, "ui"].iter().collect();
+    if path.exists() {
+        config.target_rustcflags = Some(format!(
+            "--color=never {}",
+            config.target_rustcflags.unwrap_or("".to_string())
+        ));
+        config.mode = common::Mode::Ui;
+        config.src_base = path;
+        run_tests(&config);
+    }
+
+    let path: PathBuf = ["tests", group_name, "pass"].iter().collect();
+    if path.exists() {
+        config.mode = common::Mode::RunPass;
+        config.src_base = path;
+        run_tests(&config);
+    }
+
+    let path: PathBuf = ["tests", group_name, "fail"].iter().collect();
+    if path.exists() {
+        config.mode = common::Mode::CompileFail;
+        config.src_base = path;
+        run_tests(&config);
+    }
+}
+
+fn run_no_verification(group_name: &str, filter: &Option<String>) {
+    // Automatically restored to their original value at the end of the lexical scope
```

Comment:
Seems like a comment for the `TemporaryEnvVar` struct itself, not places where it is used.

Reply:
Nice catch, thanks

Reply:
:+1: